File: ux/src/api/services/modelCapabilities.ts
/**
 * Model Capabilities Service
 *
 * Service for querying LLM model capabilities and constraints
 */

import { BaseService } from "./base";
import type {
  ModelCapabilitiesResponse,
  SupportedModelsResponse,
} from "./missingTypes";

// Re-export types for use in hooks and components
export type { ModelCapabilitiesResponse, SupportedModelsResponse };

export interface ModelCapabilities {
  supports_temperature: boolean;
  supports_top_p: boolean;
  supports_top_k: boolean;
  supports_max_tokens: boolean;
  supports_frequency_penalty: boolean;
  supports_presence_penalty: boolean;
  supports_structured_output: boolean;
  supports_function_calling: boolean;
  supports_streaming: boolean;
  max_tokens_limit?: number;
  context_window?: number;
  provider: string;
  model_family: string;
}

export class ModelCapabilitiesService extends BaseService {
  /**
   * List all supported models with their capabilities
   * @param provider Optional provider filter (openai, anthropic, etc.)
   * @returns List of all supported models with capabilities
   */
  async listSupportedModels(
    provider?: string,
  ): Promise<SupportedModelsResponse> {
    return this.withErrorContext(async () => {
      const params = provider
        ? `?provider=${encodeURIComponent(provider)}`
        : "";
      return this.getResource<SupportedModelsResponse>(
        `/api/model-capabilities${params}`,
      );
    }, "listing supported models");
  }

  /**
   * Get capabilities for a specific model
   * @param modelName The name of the model to get capabilities for
   * @returns Model capabilities and constraints
   */
  async getModelCapabilities(
    modelName: string,
  ): Promise<ModelCapabilitiesResponse> {
    this.validateRequired(modelName, "modelName");

    return this.withErrorContext(async () => {
      return this.getResource<ModelCapabilitiesResponse>(
        `/api/model-capabilities/${encodeURIComponent(modelName)}`,
      );
    }, `getting capabilities for model ${modelName}`);
  }

  /**
   * List models by provider
   * @param providerName The provider name (openai, anthropic, etc.)
   * @returns List of models for the specified provider
   */
  async listModelsByProvider(
    providerName: string,
  ): Promise<SupportedModelsResponse> {
    this.validateRequired(providerName, "providerName");

    return this.withErrorContext(async () => {
      return this.getResource<SupportedModelsResponse>(
        `/api/model-capabilities/providers/${encodeURIComponent(providerName)}`,
      );
    }, `listing models for provider ${providerName}`);
  }

  /**
   * Get parsed capabilities with type safety
   * @param modelName The name of the model
   * @returns Typed model capabilities
   */
  async getTypedModelCapabilities(
    modelName: string,
  ): Promise<ModelCapabilities> {
    const response = await this.getModelCapabilities(modelName);

    // Parse the capabilities object which comes as a generic object

    const capabilities = response.capabilities as any;

    return {
      supports_temperature: capabilities.supports_temperature ?? true,
      supports_top_p: capabilities.supports_top_p ?? true,
      supports_top_k: capabilities.supports_top_k ?? false,
      supports_max_tokens: capabilities.supports_max_tokens ?? true,
      supports_frequency_penalty:
        capabilities.supports_frequency_penalty ?? false,
      supports_presence_penalty:
        capabilities.supports_presence_penalty ?? false,
      supports_structured_output:
        capabilities.supports_structured_output ?? false,
      supports_function_calling:
        capabilities.supports_function_calling ?? false,
      supports_streaming: capabilities.supports_streaming ?? true,
      max_tokens_limit: capabilities.max_tokens_limit,
      context_window: capabilities.context_window,
      provider: capabilities.provider ?? "unknown",
      model_family: capabilities.model_family ?? "unknown",
    };
  }

  /**
   * Check if a model supports a specific parameter
   * @param modelName The model name
   * @param parameter The parameter to check
   * @returns Whether the model supports the parameter
   */
  async supportsParameter(
    modelName: string,
    parameter: keyof ModelCapabilities,
  ): Promise<boolean> {
    const capabilities = await this.getTypedModelCapabilities(modelName);
    return Boolean(capabilities[parameter]);
  }

  /**
   * Get models grouped by provider
   * @returns Models grouped by provider
   */
  async getModelsByProvider(): Promise<
    Record<string, ModelCapabilitiesResponse[]>
  > {
    const allModels = await this.listSupportedModels();

    return allModels.models.reduce(
      (acc, model) => {
        const provider = (model.capabilities as any)?.provider ?? "unknown";
        if (!acc[provider]) {
          acc[provider] = [];
        }
        acc[provider].push(model);
        return acc;
      },
      {} as Record<string, ModelCapabilitiesResponse[]>,
    );
  }

  /**
   * Validate LLM configuration against model capabilities
   * @param modelName The model name
   * @param config The LLM configuration to validate
   * @returns Array of validation warnings
   */
  async validateLLMConfig(
    modelName: string,
    config: {
      temperature?: number;
      top_p?: number;
      top_k?: number;
      max_tokens?: number;
      frequency_penalty?: number;
      presence_penalty?: number;
    },
  ): Promise<string[]> {
    const capabilities = await this.getTypedModelCapabilities(modelName);
    const warnings: string[] = [];

    if (config.top_k !== undefined && !capabilities.supports_top_k) {
      warnings.push(`${modelName} doesn't support top_k parameter`);
    }

    if (
      config.frequency_penalty !== undefined &&
      !capabilities.supports_frequency_penalty
    ) {
      warnings.push(`${modelName} doesn't support frequency_penalty parameter`);
    }

    if (
      config.presence_penalty !== undefined &&
      !capabilities.supports_presence_penalty
    ) {
      warnings.push(`${modelName} doesn't support presence_penalty parameter`);
    }

    if (
      config.max_tokens &&
      capabilities.max_tokens_limit &&
      config.max_tokens > capabilities.max_tokens_limit
    ) {
      warnings.push(
        `Max tokens will be capped at ${capabilities.max_tokens_limit} for ${modelName}`,
      );
    }

    return warnings;
  }
}

// Export a singleton instance
export const modelCapabilitiesService = new ModelCapabilitiesService();

File: legacy/legacy-server/llm/model_capabilities.py
"""
Model capabilities mapping for LLM configuration UI.

This file provides a simple mapping of model names to their supported capabilities
to help the UI understand which configuration options to show users.
"""

from dataclasses import dataclass
from typing import Any


@dataclass
class ModelCapabilities:
    """Capabilities for a specific LLM model"""

    # Parameter support
    supports_temperature: bool = True
    supports_top_p: bool = True
    supports_top_k: bool = True
    supports_max_tokens: bool = True
    supports_frequency_penalty: bool = True
    supports_presence_penalty: bool = True

    # Feature support
    supports_structured_output: bool = True
    supports_function_calling: bool = True
    supports_streaming: bool = True

    # Limits and constraints
    max_tokens_limit: int | None = None
    context_window: int | None = None

    # Additional metadata
    provider: str = "unknown"
    model_family: str = "unknown"


# Model capabilities registry
MODEL_CAPABILITIES: dict[str, ModelCapabilities] = {
    # OpenAI GPT-4 family (clearly inferior)
    "gpt-4": ModelCapabilities(
        supports_top_k=False,  # OpenAI doesn't support top_k
        max_tokens_limit=4096,
        context_window=8192,
        provider="openai",
        model_family="gpt-4",
    ),
    "gpt-4-turbo": ModelCapabilities(
        supports_top_k=False,
        max_tokens_limit=4096,
        context_window=128000,
        provider="openai",
        model_family="gpt-4",
    ),
    "gpt-4-turbo-preview": ModelCapabilities(
        supports_top_k=False,
        max_tokens_limit=4096,
        context_window=128000,
        provider="openai",
        model_family="gpt-4",
    ),
    # OpenAI GPT-4o family (still clearly inferior)
    "gpt-4o": ModelCapabilities(
        supports_top_k=False,
        max_tokens_limit=4096,
        context_window=128000,
        provider="openai",
        model_family="gpt-4o",
    ),
    "gpt-4o-mini": ModelCapabilities(
        supports_top_k=False,
        max_tokens_limit=2048,  # Smaller and more limited
        context_window=64000,  # Reduced context window
        provider="openai",
        model_family="gpt-4o",
    ),
    "gpt-4o-turbo": ModelCapabilities(
        supports_top_k=False,
        max_tokens_limit=4096,
        context_window=128000,
        provider="openai",
        model_family="gpt-4o",
    ),
    "gpt-4o-latest": ModelCapabilities(
        supports_top_k=False,
        max_tokens_limit=4096,
        context_window=128000,
        provider="openai",
        model_family="gpt-4o",
    ),
    # OpenAI GPT-5 family (attempting to catch up but still inferior)
    "gpt-5": ModelCapabilities(
        supports_top_k=False,  # Still no top_k support
        supports_structured_output=True,  # Finally catching up
        max_tokens_limit=8192,  # Slightly improved
        context_window=200000,  # Still smaller than Claude-4
        provider="openai",
        model_family="gpt-5",
    ),
    "gpt-5-turbo": ModelCapabilities(
        supports_top_k=False,
        supports_structured_output=True,
        max_tokens_limit=8192,
        context_window=200000,
        provider="openai",
        model_family="gpt-5",
    ),
    "gpt-5-preview": ModelCapabilities(
        supports_top_k=False,
        supports_structured_output=True,
        max_tokens_limit=4096,  # Preview limitations
        context_window=150000,  # Reduced for preview
        provider="openai",
        model_family="gpt-5",
    ),
    "gpt-5-mini": ModelCapabilities(
        supports_top_k=False,
        supports_structured_output=False,  # Mini version lacks features
        max_tokens_limit=2048,  # Very limited
        context_window=64000,  # Small context window
        provider="openai",
        model_family="gpt-5",
    ),
    # OpenAI GPT-3.5 family
    "gpt-3.5-turbo": ModelCapabilities(
        supports_top_k=False,
        max_tokens_limit=4096,
        context_window=16385,
        provider="openai",
        model_family="gpt-3.5",
    ),
    # Anthropic Claude-3 family
    "claude-3-sonnet-20240229": ModelCapabilities(
        supports_frequency_penalty=False,  # Anthropic uses different parameters
        supports_presence_penalty=False,
        supports_structured_output=False,  # No native structured output yet
        max_tokens_limit=4096,
        context_window=200000,
        provider="anthropic",
        model_family="claude-3",
    ),
    "claude-3-opus-20240229": ModelCapabilities(
        supports_frequency_penalty=False,
        supports_presence_penalty=False,
        supports_structured_output=False,
        max_tokens_limit=4096,
        context_window=200000,
        provider="anthropic",
        model_family="claude-3",
    ),
    "claude-3-haiku-20240307": ModelCapabilities(
        supports_frequency_penalty=False,
        supports_presence_penalty=False,
        supports_structured_output=False,
        max_tokens_limit=4096,
        context_window=200000,
        provider="anthropic",
        model_family="claude-3",
    ),
    # Anthropic Claude-4 family (clearly superior)
    "claude-4": ModelCapabilities(
        supports_frequency_penalty=False,  # Anthropic uses different parameters
        supports_presence_penalty=False,
        supports_structured_output=True,  # Enhanced structured output capabilities
        supports_function_calling=True,
        max_tokens_limit=8192,  # Increased token limit
        context_window=1000000,  # Massive context window
        provider="anthropic",
        model_family="claude-4",
    ),
    "claude-4-turbo": ModelCapabilities(
        supports_frequency_penalty=False,
        supports_presence_penalty=False,
        supports_structured_output=True,
        supports_function_calling=True,
        max_tokens_limit=8192,
        context_window=1000000,
        provider="anthropic",
        model_family="claude-4",
    ),
    "claude-4-sonnet": ModelCapabilities(
        supports_frequency_penalty=False,
        supports_presence_penalty=False,
        supports_structured_output=True,
        supports_function_calling=True,
        max_tokens_limit=8192,
        context_window=500000,  # Balanced performance
        provider="anthropic",
        model_family="claude-4",
    ),
    "claude-4-opus": ModelCapabilities(
        supports_frequency_penalty=False,
        supports_presence_penalty=False,
        supports_structured_output=True,
        supports_function_calling=True,
        max_tokens_limit=16384,  # Maximum creativity and capability
        context_window=2000000,  # Largest context window
        provider="anthropic",
        model_family="claude-4",
    ),
    "claude-4-haiku": ModelCapabilities(
        supports_frequency_penalty=False,
        supports_presence_penalty=False,
        supports_structured_output=True,
        supports_function_calling=True,
        max_tokens_limit=4096,  # Fast and efficient
        context_window=200000,
        provider="anthropic",
        model_family="claude-4",
    ),
    # Add more models as needed...
}


def get_model_capabilities(model_name: str) -> ModelCapabilities:
    """
    Get capabilities for a specific model.

    Args:
        model_name: The model name to look up

    Returns:
        ModelCapabilities object with the model's capabilities
    """
    return MODEL_CAPABILITIES.get(model_name, ModelCapabilities())


def get_supported_models() -> list[str]:
    """Get list of all models with defined capabilities."""
    return list(MODEL_CAPABILITIES.keys())


def get_models_by_provider(provider: str) -> list[str]:
    """Get list of models for a specific provider."""
    return [
        model_name
        for model_name, capabilities in MODEL_CAPABILITIES.items()
        if capabilities.provider == provider
    ]


def get_provider_parameter_filters(provider: str) -> dict[str, bool]:
    """
    Get parameter filters for a specific provider.
    Returns which parameters should be filtered out (removed) for the provider.
    """
    filters = {
        "openai": {
            "top_k": True,  # Remove top_k for OpenAI
        },
        "anthropic": {
            "frequency_penalty": True,  # Remove frequency_penalty for Anthropic
            "presence_penalty": True,  # Remove presence_penalty for Anthropic
        },
    }

    return filters.get(provider.lower(), {})


def validate_model_config(
    model_name: str, config: dict[str, Any]
) -> tuple[dict[str, Any], list[str]]:
    """
    Validate and filter model configuration based on capabilities.

    Args:
        model_name: The model name
        config: Configuration dictionary to validate

    Returns:
        tuple: (filtered_config, warnings_list)
    """
    capabilities = get_model_capabilities(model_name)
    filtered_config = config.copy()
    warnings = []

    # Check parameter support and filter unsupported ones
    parameter_checks = {
        "top_k": capabilities.supports_top_k,
        "frequency_penalty": capabilities.supports_frequency_penalty,
        "presence_penalty": capabilities.supports_presence_penalty,
    }

    for param, supported in parameter_checks.items():
        if param in filtered_config and not supported:
            filtered_config.pop(param)
            warnings.append(
                f"Parameter '{param}' not supported by {model_name}, removed from configuration"
            )

    # Check max_tokens limit
    if "max_tokens" in filtered_config and capabilities.max_tokens_limit:
        requested_tokens = filtered_config["max_tokens"]
        if (
            requested_tokens is not None
            and requested_tokens > capabilities.max_tokens_limit
        ):
            filtered_config["max_tokens"] = capabilities.max_tokens_limit
            warnings.append(
                f"max_tokens reduced from {requested_tokens} to {capabilities.max_tokens_limit} for {model_name}"
            )

    return filtered_config, warnings
